File: medianOfSorted.py
#Time Complexity : O(log(m)) - Smaller array
#Space Complexity : O(1)


class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        n1 = len(nums1)
        n2 = len(nums2)

        if n1 > n2:
            return self.findMedianSortedArrays(nums2, nums1)

        l, h = 0, n1
        
        while l<=h:
            mid = (l+h)//2
            mid2 = (n1+n2)//2 - mid
            l1 = nums1[mid-1] if mid!=0 else float("-inf")
            r1 = nums1[mid] if mid!=n1 else float("inf")
            l2 = nums2[mid2-1] if mid2!=0 else float("-inf")
            r2 = nums2[mid2] if mid2!=n2 else float("inf")
            

            if l1 <= r2 and l2 <= r1:

                if (n1+n2)%2 == 0:
                    return (max(l1, l2) + min(r1, r2))/2
                else:
                    return min(r1, r2)
            elif l2 > r1:
                l = mid + 1
            else:
                h = mid - 1
            

File: MedianOfSorted.java
public class MedianOfSorted {
    //TC = O(log m) -- BS on smaller array
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        if(nums1 == null|| nums2==null)return 0.0;
        int m = nums1.length;
        int n = nums2.length;
        if(m>n) return findMedianSortedArrays(nums2,nums1);
        int low = 0;
        int high = m;
        while(low<=high){
            int partX = low+(high-low)/2;
            int partY = (m+n)/2 - partX;
            double L1 = partX ==0 ?Integer.MIN_VALUE :nums1[partX-1];
            double R1 = partX ==m ?Integer.MAX_VALUE :nums1[partX];
            double L2 = partY ==0 ?Integer.MIN_VALUE :nums2[partY-1];
            double R2 = partY ==n ?Integer.MAX_VALUE :nums2[partY];

            if(L1<=R2 && L2<=R1){
                if((m+n)%2 ==0){
                    return (Math.min(R1,R2) + Math.max(L1,L2))/2;
                }
                else{
                    return Math.min(R1,R2);
                }
            }
            else if(L1>R2){
                high = partX-1;
            }
            else{
                low = partX+1;
            }

        }
        return 9.00;
    }

}
